codecov_auth/views/okta.py

```diff
@@ -0,0 +1,191 @@
+        id_token,
+        key=key,
+        algorithms=["RS256"],
+        audience=settings.OKTA_OAUTH_CLIENT_ID,
+    )
+    assert id_payload["iss"] == iss
+    assert id_payload["aud"] == settings.OKTA_OAUTH_CLIENT_ID
+
+    return id_payload
+
+
+auth = HTTPBasicAuth(settings.OKTA_OAUTH_CLIENT_ID, settings.OKTA_OAUTH_CLIENT_SECRET)
+
+
+class OktaLoginView(LoginMixin, View):
+    def _fetch_user_data(self, iss: str, code: str) -> Optional[Dict]:
+
+        res = requests.post(
+            f"{iss}/oauth2/v1/token",
+            auth=auth,
+            data={
+                "grant_type": "authorization_code",
+                "code": code,
+                "redirect_uri": settings.OKTA_OAUTH_REDIRECT_URL,
+            },
+        )
+
+        if res.status_code >= 400:
+            return None
+        return res.json()
+
+    def _redirect_to_consent(self, iss: str) -> HttpResponse:
+        qs = urlencode(
+            dict(
+                response_type="code",
+                client_id=settings.OKTA_OAUTH_CLIENT_ID,
+                scope="openid email profile",
+                redirect_uri=settings.OKTA_OAUTH_REDIRECT_URL,
+                state=iss,
+            )
+        )
+        redirect_url = f"{iss}/oauth2/v1/authorize?{qs}"
+        response = redirect(redirect_url)
+        self.store_to_cookie_utm_tags(response)
+
+        if settings.OKTA_ISS is None:
+            # ISS was passed to us from Okta
+            self._store_iss_cookie(iss, response)
+
+        return response
+
+    def _store_iss_cookie(self, iss: str, response: HttpResponse):
+        response.set_cookie(
+            "_okta_iss",
+            iss,
+            max_age=86400,  # Same as state validatiy
+            httponly=True,
+            domain=settings.COOKIES_DOMAIN,
+        )
+
+    def _perform_login(self, request: HttpRequest) -> HttpResponse:
+        code = request.GET.get("code")
+        iss = settings.OKTA_ISS or request.COOKIES.get("_okta_iss")
+        if iss is None:
+            log.warning("Unable to log in due to missing Okta issuer", exc_info=True)
+            return redirect(f"{settings.CODECOV_DASHBOARD_URL}/login")
+
+        user_data = self._fetch_user_data(iss, code)
+        if user_data is None:
+            log.warning("Unable to log in due to problem on Okta", exc_info=True)
+            return redirect(f"{settings.CODECOV_DASHBOARD_URL}/login")
+
+        current_user = self._login_user(request, iss, user_data)
+
+        # TEMPORARY: we're assuming a single owner for the time being since there's
+        # no supporting UI to select which owner you'd like to view
+        owner = current_user.owners.first()
+        if owner is not None:
+            service = get_short_service_name(owner.service)
+            response = redirect(f"{settings.CODECOV_DASHBOARD_URL}/{service}")
+        else:
+            # user has not connected any owners yet
+            response = redirect(f"{settings.CODECOV_DASHBOARD_URL}/sync")
```

Comment:
Is this a new url? I don't think I've seen this sync endpoint 👀

Reply:
Yeah a new page that Nick created to support "Login with Sentry".  Once a user is authenticated but doesn't have any linked Git providers we send them to this page (where they can sync a Git provider).

Reply:
Oh it's not merged yet: https://github.com/codecov/gazebo/pull/2235